src/ast.cpp

```diff
@@ -48,11 +48,90 @@
 
 using namespace ispc;
 
+///////////////////////////////////////////////////////////////////////////
+// Indent
+
+Indent::~Indent() {
+    Assert(stack.empty() && "Indent stack is not empty on destruction");
+    // Any Print() call must be paired by Done() call
+    Assert(printCalls == doneCalls && "AST dump has encountered a bug");
+};
+
+void Indent::pushSingle() { stack.push_back(1); }
+void Indent::pushList(int i) {
+    if (i > 0) {
+        stack.push_back(i);
+    }
+}
+
+void Indent::setNextLabel(std::string s) { label = s; }
+
+void Indent::Print() {
+    printCalls++;
+    Assert(!stack.empty());
+    int &top = stack.back();
+    Assert(top > 0);
+
+    for (int i = 0; i < (stack.size() - 1); i++) {
+        if (stack[i] == 0) {
+            printf("  ");
+        } else {
+            printf("| ");
+        }
+    }
+
+    if (top == 1) {
+        printf("`-");
+    } else {
+        printf("|-");
+    }
+    top--;
+
+    if (!label.empty()) {
+        printf("(%s) ", label.c_str());
+        label.clear();
+    }
+}
+
+void Indent::Print(const char *title) {
+    Print();
+    if (title != nullptr) {
+        printf("%s", title);
+    }
+}
+
+void Indent::Print(const char *title, const SourcePos &pos) {
+    Print(title);
+    pos.Print();
+}
+
+void Indent::PrintLn(const char *title, const SourcePos &pos) {
+    Print(title, pos);
+    printf("\n");
+}
```

Comment:
Why `printf()` instead of `Print()`?

Reply:
`Indent::Print()` functions are implemented using `printf()`.

`PrintLn()` does the same as `Print()`, but prints `\n` in the on top of it.